src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -231,11 +249,153 @@ def _estimate_third_rows(self, vijs, viis):
+        all_shift_phases = shift_phases.T
+
+        n_pairs = n_img * (n_img - 1) // 2
+        max_corrs = np.zeros(n_pairs)
+        max_corrs_idx = np.zeros(n_pairs)
+
+        # Q is the n_img x n_img  Hermitian matrix defined by Q = q*q^H,
+        # where q = (exp(i*order*theta_0), ..., exp(i*order*theta_{n_img-1}))^H,
+        # and theta_i in [0, 2pi/order) is the in-plane rotation angle for the i'th image.
+        Q = np.zeros((n_img, n_img), dtype=complex)
+
+        # Transpose pf and reconstruct the full polar Fourier for use in correlation.
+        # self.pf only consists of rays in the range [180, 360).
+        pf = pf.transpose((2, 1, 0))
+        pf = np.concatenate((pf, np.conj(pf)), axis=1)
+
+        # Normalize rays.
+        pf /= norm(pf, axis=-1)[..., np.newaxis]
+
+        with tqdm(total=n_pairs) as pbar:
+            idx = 0
+            # Note: the ordering of i and j in these loops should not be changed as
+            # they correspond to the ordered tuples (i, j), for i<j.
+            for i in range(n_img):
+                pf_i = pf[i]
+
+                # Generate shifted versions of images.
+                pf_i_shifted = np.array(
+                    [pf_i * shift_phase for shift_phase in all_shift_phases]
+                )
+
+                Ri_tilde = Ri_tildes[i]
+
+                for j in range(i + 1, n_img):
+                    pf_j = pf[j]
+
+                    Rj_tilde = Ri_tildes[j]
+
+                    # Compute all possible rotations between the i'th and j'th images.
+                    Us = np.array(
+                        [
+                            Ri_tilde.T @ R_theta_ij @ Rj_tilde
+                            for R_theta_ij in R_theta_ijs
+                        ]
+                    )
+
+                    # Find the angle between common lines induces by the rotations.
+                    c1s = np.array([[-U[1, 2], U[0, 2]] for U in Us])
+                    c2s = np.array([[U[2, 1], -U[2, 0]] for U in Us])
+
+                    # Convert from angles to indices.
+                    c1s = self.clAngles2Ind(c1s, n_theta)
+                    c2s = self.clAngles2Ind(c2s, n_theta)
+
+                    # Perform correlation.
+                    corrs = np.array(
+                        [
+                            np.dot(pf_i_shift[c1], np.conj(pf_j[c2]))
+                            for pf_i_shift in pf_i_shifted
+                            for c1, c2 in zip(c1s, c2s)
+                        ]
+                    )
+
+                    # Reshape to group by shift and symmetric order.
+                    corrs = corrs.reshape((n_shifts, order, len(theta_ijs) // order))
+
+                    # For each pair of lines we take the maximum correlation over all shifts.
+                    corrs = np.max(np.real(corrs), axis=0)
+
+                    # We take the mean score over the 'order' groups and find the group that attains the maximum.
```

Comment:
What does this mean?

Reply:
Sorry this was a little confusing. At this point `corrs` has shape `(order, len(theta_ijs) // order)`, where `corrs[i]` is the correlation over all in-plane rotations in the range `[2pi * i / order, 2pi * (i + 1) / order)`. Due to the symmetry of the molecule each set `corrs[i]` represent the same collection of correlations, so we average them together before finding the max.

I will make some clarifying comments.

Reply:
Comments added.